src/DensityTest.jl

```diff
@@ -0,0 +1,272 @@
+            
+            
+    rf = 2 * rcoef[3] .+ 6 * rcoef[4] * rdf.X + 12 * rcoef[5] * rdf.X.^2 
+    rh = κ * (rmse * (rdf.X[end] - c) / sum(rf'rf))^(0.2)
+            
+        # Taking the average
+    h = 0.5 * (rh + lh)
+end
+
+# 6. McCraryTest
+"""
+    fit(method::McCraryTest, data::RunningVariable)
+
+Test for manipulation in the running variable following McCrary (2008).
+
+# Arguments
+
+- `data::RunningVariable`: The running variable.
+
+# Returns
+Returns  a FittedMcCraryTest object with fields `b` and `h` which are the bin size and the bandwidth used, respectively. Also, it returns `θhat`, `σθ`, `pval`, `Js`, and `fitted_res` which are the estimated difference in the intercepts, the standard error of the estimate, the pval of the test H0: θhat = 0, the number of bins used, and the fitted values for the support of the running variable (as a DataFrame). Finally, `coeftable` provides an object of type CoefTable for the estimator. 
+"""
+function fit(method::McCraryTest, data::RunningVariable)
+    
+    # Cutoff selection
+    c = data.cutoff
+    R = data.Zs
+    # Call the functions 
+    # 1.
+    n, sd, rmax, rmin = DescriptiveStats(R)
+
+    # 2.
+    if method.bin === nothing
+        b = Binsize(n, sd)
+    else
+        b = method.bin
+    end
+
+    # 3.
+    J, Jl, Jr = BinNumber(c, b, rmin, rmax)
+    Js = [J; Jl; Jr]
+
+    # 4.
+    X, Y = DRV(R, c, b, rmin, J, n)
+
+    # 5.
+    if method.bw === nothing
+        h = McCraryBandwidth(X, Y, Jl, c)
+    else
+        h = method.bw
+    end
+    
+    ##############################################################
+    # Smooth histograms to the left and to the right of the cutoff
+    ldf = DataFrame(Y=Y[1:Jl], X=X[1:Jl], pred=zeros(Jl))
+    rdf = DataFrame(Y=Y[Jl + 1:end], X=X[Jl + 1:end], pred=zeros(Jr))
+
+    # Left side 
+    Sminus(k) = sum(TriangularKernel.((ldf.X .- c) ./ h) .* (ldf.X .- c).^k)
+    lfhat = sum(TriangularKernel.((ldf.X .- c) ./ h) .* ((Sminus(2) .- Sminus(1) .* (ldf.X .- c)) ./ (Sminus(2) * Sminus(0) - (Sminus(1))^2 )) .* ldf.Y)
+ 
+    # Right side
+    Splus(k) = sum(TriangularKernel.((rdf.X .- c) ./ h) .* (rdf.X .- c).^k)
+    rfhat = sum(TriangularKernel.((rdf.X .- c) ./ h) .* ((Splus(2) .- Splus(1) .* (rdf.X .- c)) ./ (Splus(2) * Splus(0) - (Splus(1))^2)) .* rdf.Y)
+ 
+    # The estimates:
+    θhat = log(rfhat) - log(lfhat)
+    σθ = sqrt((1 / (n * h)) * (24 / 5) * ( (1 / rfhat) + (1 / lfhat)))
+    z = θhat / σθ
+    pval = 2(1 -  cdf(Normal(), abs(z)))
+
+    # Creating the table
+    res = [θhat σθ z pval]
+    colnms = ["θ̂"; "σ̂"; "z"; "p-val"]
+    rownms = ["McCrary Test"]
+    coeftbl = CoefTable(res, colnms, rownms, 4, 3)
+
+    # Fitted Values:
+        # Left side
+    for i in eachindex(ldf.X)
+        ldf[!,:dist] = ldf.X  .-  ldf.X[i]
+        lwght = TriangularKernel.(ldf.dist / h)
+        laux = lm(@formula(Y ~ dist), ldf, wts=lwght)
+        ldf.pred[i] = predict(laux)[i]
+    end
+
+    m = min.((c .- ldf.X) / h, 1)
+
```

Comment:
What is `lVarf`?

Reply:
`lVarf` and `rVarf` are the variances of the estimates to the left and to the right of the cutoff `(V[f(r)])`. It comes from equation (4) in the supplementary material of the paper (attached).

[DCdensity.pdf](https://github.com/nignatiadis/RegressionDiscontinuity.jl/files/5923121/DCdensity.pdf)